### PR TITLE
[GitHub repositoryQuery] Retry searches with missed results

### PR DESCRIPTION
**Pull request by pjlast**
While iterating over a search result, because of GitHub search inconsistencies, we can miss some results.

But, because we know how many results we're supposed to get, thanks to the `repositoryCount` field, we can retry the request again until we've seen all the repositories.

Missing results should not happen with every search, and we won't have to iterate over ALL the results to find the missing ones. As soon as we've seen enough, we immediately break.

We also put a limit on the number of times we can iterate over a search.

## Test plan

Fundamentally the same.

### CODE REVIEW

**Comment by varsanojidan** [comment]
Is this to cover the situation in which more than 1000 repos are created in the same second in which case we would miss some results? 

Are there other scenarios we can miss repos?

File: internal/repos/github.go [hunk 1]
```diff
@@ -1056,14 +1058,32 @@ func (q *repositoryQuery) doRecursively(ctx context.Context, results chan *githu
...
+		// Only break if we've seen a number of repositories equal to the expected count
+		// res.EndCursor will loop by itself
+		if len(seen) >= res.TotalCount {
 			break
 		}
+
+		// Set a hard cap on the number of retries
+		if res.EndCursor == "" {
+			numTries += 1
+			if numTries >= 3 {
```
Comment (vdavid) [inline internal/repos/github.go:1082]: Extract magic value to a constant?

**Review by sashaostrikov — APPROVED** [review]
LGTM!
How hard can it be to write a test for this thing?

**Comment by vdavid** [comment]
@pjlast Not strictly related to this PR, but for `doRecursively`, we do "If the number of search results returned is greater than the query limit, split the query in half by filtering by repo creation date, and perform those two queries. Do so recursively." → why two? Shouldn't the query be split into `ceil(q.RepoCount.count / q.Limit)` queries right away?

**Comment by varsanojidan** [comment]
>@varsanojidan its to fix cases like a repository gaining stars while paginating through the results, thus throwing off the page ordering, which causes us to see some repos multiple times and completely miss other repos

Ahh gotcha, makes sense :)

**Review by varsanojidan — APPROVED** [review]

**Comment by pjlast** [comment]
@vdavid we don't know how the repositories are distributed over time. 7000 repositories could have been created on one day, and 100 repositories on the next day. We cannot assume that the results are evenly distributed over time.

**Pull request merged**